dmutils/timing.py

```diff
@@ -0,0 +1,91 @@
+from contextlib import contextmanager
+import logging
+import sys
+from time import perf_counter, process_time
+
+from flask import request
+from flask.ctx import has_request_context
+
+
+#
+# logged_duration's defaults are exposed here so that a caller is able to defer to the default or simply use a wrapper
+# for the default when customizing the call.
+#
+
+
+def default_message(log_context):
+    return "Block {} in {{duration_real}}s of real-time".format(
+        "executed" if sys.exc_info()[0] is None else "raised {}".format(sys.exc_info()[0].__name__)
+    )
+
+
+def default_condition(log_context):
+    return has_request_context() and getattr(request, "sampling_decision", False)
+
+
+def default_log_func(logger, message, log_level, log_context):
+    final_message = message(log_context) if callable(message) else message
+    logger.log(log_level, final_message, exc_info=(sys.exc_info()[0] is not None), extra=log_context)
+
+
+# exposing this allows a caller to specify default_logger.getChild(...) as their logger
+default_logger = logging.getLogger(__name__)
+
+
+@contextmanager
+def logged_duration(
+    logger=default_logger,
+    message=default_message,
+    log_level=logging.DEBUG,
+    condition=default_condition,
+    log_func=default_log_func,
+):
+    """
+        returns a context manager which will monitor the amount of time spent "inside" its code block and emit a log
+        message on exiting block if ``condition`` passes. Uses a ``log_context`` dictionary as the log call's ``extra``
+        parameter which will contain the parameters ``duration_real`` and ``duration_process``, each being a float
+        duration in seconds. Additional parameters can be added to this ``log_context`` dictionary by annotating them
+        to the dictionary which is yielded by the context manager, e.g.::
+
+            with logged_duration(message="Received result {foo} in {duration_real}s") as log_context:
+                ...
+                log_context["foo"] = do_something()
+                ...
+
+        Logging action is triggered whether execution fell out of the block naturally or the code raised an exception.
+        If log messages are not desired on exeptions these can always be excluded by a custom ``condition``.
+
+        Can also be used as a function decorator.
+
+        :param logger:    The logger to log the message to. It's best to set this at the very least to give the log
+                          reader a clue as to what block of code is being timed.
+        :param message:   Message format string to emit. Can be a callable which should return the desired message
```

Comment:
Is this a standard thing, cos I find it a bit weird - wouldn't it be more clear to have two alternative parameters, i.e. `message` and `message_factory` or something? I'm not sure I really understand the use case for having a function that generates the message based a context anyway..

Reply:
Yes I've seen it quite a lot, notably in the `mock` module. When you have multiple parameters you've got to add logic around which overrides which etc. For a use case see the default log message (function). That's the case I added it for because otherwise I'd always be implying that the block succeeded.